### PR TITLE
Feat/service import backend support

### PR DESCRIPTION
**Pull request by tanujd11**
**What type of PR is this?**
 Feature

**What this PR does / why we need it**:
Adds serviceImport Backend for Routes.

**Which issue(s) this PR fixes**:

Fixes https://github.com/envoyproxy/gateway/issues/9

### CODE REVIEW

**Comment by arkodg** [comment]
thanks for picking this feature up @tanujd11 ! would be great if this land in v0.6
quick q - isn't this feature blocked on https://github.com/envoyproxy/gateway/pull/1494 which is adding EndpointSlice support ?

**Comment by tanujd11** [comment]
@arkodg , I think this can go in with using ClusterSetIP type in multi cluster service import. EndpointSlice can be used once that PR gets in. It will be similar to service backend.

**Comment by arkodg** [comment]
Rethinking, you're right @tanujd11, this PR can go in by itself.
Will review this soon once we start working on v0.6 features.

File: charts/gateway-helm/crds/multicluster-svc.yaml
Comment (arkodg) [inline charts/gateway-helm/crds/multicluster-svc.yaml:1]: hey lets rm the CRD from here, and add it as part of docs

File: internal/gatewayapi/translator_test.go
Comment (arkodg) [inline internal/gatewayapi/translator_test.go:98]: prefer if we dont add this by default here, but in the YAML testdata input, since this resource is not the common case

File: internal/provider/kubernetes/kubernetes_test.go
Comment (arkodg) [inline internal/provider/kubernetes/kubernetes_test.go:91]: this could live under `internal/provider/kubernetes/testdata/mcs-crds/`

**Comment by arkodg** [comment]
did a first pass, it looks good ! added some minor comments

**Review by Xunzhuo — COMMENTED** [review]
This is awesome, but there are some conflcts : )

It would be a nice feat in v0.6.0.

**Comment by arkodg** [comment]
ptal @envoyproxy/gateway-maintainers, a decision also needs be made here on whether we want the ServiceImport API to be opt in or not

**Comment by tanujd11** [comment]
@arkodg, if we are removing the serviceImport CRD from EG chart, then it automatically becomes the a Opt-in feature where user needs to install the CRD to use this feature. I can add a validation to check if the CRD exist before someone uses serviceImport as part of the backend. for example if the user has mentioned serviceImport as backendRef and the CRD does not exist, we can reject the config with status `backendRef kind does not exist`

**Comment by arkodg** [comment]
> @arkodg, if we are removing the serviceImport CRD from EG chart, then it automatically becomes the a Opt-in feature where user needs to install the CRD to use this feature. I can add a validation to check if the CRD exist before someone uses serviceImport as part of the backend. for example if the user has mentioned serviceImport as backendRef and the CRD does not exist, we can reject the config with status `backendRef kind does not exist`

great idea, +1 to this, but from an implementation perspective, won't the `provider` layer error out if it tries to `Watch` the `ServiceImport` resource w/o have a CRD installed ?

**Review by arkodg — APPROVED** [review]
LGTM, thanks for adding this !

**Comment by tanujd11** [comment]
> great idea, +1 to this, but from an implementation perspective, won't the `provider` layer error out if it tries to `Watch` the `ServiceImport` resource w/o have a CRD installed ?

You are right, the current one is throwing error. Should we implement a way to check the presence of CRD for watching also. Or is there any other way in your mind to tackle this

**Comment by arkodg** [comment]
@tanujd11 can you share the error you are seeing ?

one way to solve the opt in problem is add a knob in[ EnvoyGateway.ExtensionApis](https://github.com/envoyproxy/gateway/blob/b08d33579f80f0acf8c4e03febe0b39112cc611f/api/config/v1alpha1/envoygateway_types.go#L132)

**Comment by Xunzhuo** [comment]
LGTM after conflicts are resolved and CI passed.

**Comment by tanujd11** [comment]
@arkodg , can you take a relook at this work? I have added the knob at the place suggested by you. I will work on the doc in a subsequent PR

**Review by zirain — DISMISSED** [review]

**Comment by arkodg** [comment]
had a chat with @tanujd11 who is trying to implement the functionality w/o having to add an explicit API knob

**Review by Xunzhuo — DISMISSED** [review]
Looks like this PR has been rebased a couple of times, maybe we should get this in higher priority, in case more rebases. Thanks @tanujd11, mind having another rebase to fix conflicts?

File: internal/provider/kubernetes/controller.go
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:1183]: is this needed ? can we use `r.log` directly ?

File: internal/provider/kubernetes/controller.go
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:1469]: nit: ServiceImportCRDExists

File: internal/provider/kubernetes/controller.go
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:1287]: lets use `handler.EnqueueRequestsFromMapFunc(r.enqueueClass)` here instead, similar to what is done for all other `Watch` calls, might require a `rebase`, since this change was done recently

**Comment by arkodg** [comment]
hey @tanujd11 just took a peek at your PR, the approach of using the o/p. of `discoveryClient.ServerResources()` to check if the service import API exists in the API server is a great way to decide on whether to watch the resource or not.

**Comment by arkodg** [comment]
seeing some test yaml fail, recommend running `make testdata` to fix it

File: internal/provider/kubernetes/controller.go
Comment (tanujd11) [reply internal/provider/kubernetes/controller.go]: removed it, this is not required.

File: internal/provider/kubernetes/controller.go
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:1286]:
```suggestion
			r.log.Info("Unable to watch ServiceImport: %s", err.Error())
```

File: internal/provider/kubernetes/controller.go
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:1472]:
```suggestion
		r.log.Error(err, "failed to get API resource list")
```

File: internal/provider/kubernetes/controller.go
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:1468]:
```suggestion
		r.log.Error(err, "failed to create discovery client")
```

File: internal/provider/kubernetes/controller.go
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:1286]:
```suggestion
			r.log.Info("unable to watch not serviceImport: %s", err.Error())
```

Comment (arkodg) [inline internal/provider/kubernetes/controller.go:1286]:
```suggestion
			r.log.Info("unable to watch ServiceImport: %s", err.Error())
```

**Review by arkodg — APPROVED** [review]
LGTM, added some minor nits around error string
thanks for adding this !

**Comment by zirain** [comment]
can you add a user face doc in following PR.

**Review by arkodg — APPROVED** [review]

**Pull request merged**
